include/boost/math/ccmath/signbit.hpp

```diff
@@ -7,19 +7,173 @@
 
 namespace detail {
 
+#ifdef BOOST_MATH_BIT_CAST
+
+struct IEEEf2bits
+{
+#if BOOST_MATH_ENDIAN_LITTLE_BYTE
+    std::uint32_t mantissa : 23;
+    std::uint32_t exponent : 8;
+    std::uint32_t sign : 1;
+#else // Big endian
+    std::uint32_t sign : 1;
+    std::uint32_t exponent : 8;
+    std::uint32_t mantissa : 23;
+#endif 
+};
+
+struct IEEEd2bits
+{
+#if BOOST_MATH_ENDIAN_LITTLE_BYTE
+    std::uint32_t mantissa_l : 32;
+    std::uint32_t mantissa_h : 20;
+    std::uint32_t exponent : 11;
+    std::uint32_t sign : 1;
+#else // Big endian
+    std::uint32_t sign : 1;
+    std::uint32_t exponent : 11;
+    std::uint32_t mantissa_h : 20;
+    std::uint32_t mantissa_l : 32;
+#endif
+};
+
+// 80 bit long double
+#if LDBL_MANT_DIG == 64 && LDBL_MAX_EXP == 16384
+
+struct IEEEl2bits
+{
+#if BOOST_MATH_ENDIAN_LITTLE_BYTE
+    std::uint32_t mantissa_l : 32;
+    std::uint32_t mantissa_h : 32;
+    std::uint32_t exponent : 15;
+    std::uint32_t sign : 1;
+    std::uint32_t pad : 32;
+#else // Big endian
+    std::uint32_t pad : 32;
+    std::uint32_t sign : 1;
+    std::uint32_t exponent : 15;
+    std::uint32_t mantissa_h : 32;
+    std::uint32_t mantissa_l : 32;
+#endif
+};
+
+// 128 bit long double
+#elif LDBL_MANT_DIG == 113 && LDBL_MAX_EXP == 16384
+
+struct IEEEl2bits
+{
+#if BOOST_MATH_ENDIAN_LITTLE_BYTE
+    std::uint64_t mantissa_l : 64;
+    std::uint64_t mantissa_h : 48;
+    std::uint32_t exponent : 15;
+    std::uint32_t sign : 1;
+#else // Big endian
+    std::uint32_t sign : 1;
+    std::uint32_t exponent : 15;
+    std::uint64_t mantissa_h : 48;
+    std::uint64_t mantissa_l : 64;
+#endif
+};
+
+// 64 bit long double (double == long double on ARM)
+#elif LDBL_MANT_DIG == 53 && LDBL_MAX_EXP == 1024
+
+struct IEEEl2bits
+{
+#if BOOST_MATH_ENDIAN_LITTLE_BYTE
+    std::uint32_t mantissa_l : 32;
+    std::uint32_t mantissa_h : 20;
+    std::uint32_t exponent : 11;
+    std::uint32_t sign : 1;
+#else // Big endian
+    std::uint32_t sign : 1;
+    std::uint32_t exponent : 11;
+    std::uint32_t mantissa_h : 20;
+    std::uint32_t mantissa_l : 32;
+#endif
+};
+
+#else // Unsupported long double representation
+#  define BOOST_MATH_UNSUPPORTED_LONG_DOUBLE
+#endif
+
+template <typename T>
+constexpr bool signbit_impl(T arg)
+{
+    if constexpr (std::is_same_v<T, float>)
+    {   
+        const auto u = BOOST_MATH_BIT_CAST(IEEEf2bits, arg);
+        return u.sign;
+    }
+    else if constexpr (std::is_same_v<T, double>)
+    {
+        const auto u = BOOST_MATH_BIT_CAST(IEEEd2bits, arg);
+        return u.sign;
+    }
+    #ifndef BOOST_MATH_UNSUPPORTED_LONG_DOUBLE
+    else if constexpr (std::is_same_v<T, long double>)
+    {
+        const auto u = BOOST_MATH_BIT_CAST(IEEEl2bits, arg);
+        return u.sign;
+    }
+    #endif
+    else
+    {
+        if (boost::math::ccmath::isnan(arg))
+        {
+            return false;
+        }
+        
+        return arg < static_cast<T>(0);
+    }
+}
+
+#else
+
```

Comment:
If bit_cast is unavailable, would it be better to static_assert if the argument is zero or NaN on the grounds that we just don't know what the result should be?

Reply:
I think that is fair, and also matches what we discussed on the ML.